src/abi_arm.cpp

```diff
@@ -0,0 +1,315 @@
+    // so use that definition of legality (section 6.1.2.1)
+
+    if (jl_is_structtype(dt)) {
+        // ... with one to four Elements.
+        size_t members = jl_datatype_nfields(dt);
+        if (members < 1 || members > 4)
+            return 0;
+
+        base = NULL;
+        for (size_t i = 0; i < members; ++i) {
+            Type *T = isLegalHAType((jl_datatype_t*)jl_field_type(dt,i));
+            if (!T)
+                return 0;
+
+            if (!base)
+                base = T;
+            else if (base != T)
+                return 0;
+        }
+        return members;
+    }
+
+    return 0;
+}
+
+// Determine if an argument can be passed through a coprocessor register.
+//
+// All the out parameters should be default to `false`.
+static void classify_cprc(jl_datatype_t *dt, bool *vfp)
+{
+    // Based on section 6.1 of the Procedure Call Standard
+
+    // VFP: 6.1.2.1
+    // - A half-precision floating-point type.
+    // - A single-precision floating-point type.
+    // - A double-precision floating-point type.
+    if (get_llvm_fptype(dt)) {
+        *vfp = true;
+        return;
+    }
+
+    // - A 64-bit or 128-bit containerized vector type.
+    size_t vec_size = isContainerizedVector(dt);
+    if (vec_size == 8 || vec_size == 16) {
+        *vfp = true;
+        return;
+    }
+
+    // - A Homogeneous Aggregate
+    Type *base = NULL;
+    if (isLegalHA(dt, base)) {
+        *vfp = true;
+        return;
+    }
+}
+
+static void classify_return_arg(jl_value_t *ty, bool *reg, bool *vfpreg,
+                                bool *onstack, bool *need_rewrite)
+{
+    // Assume jl_is_datatype(ty) && !jl_is_abstracttype(ty)
+    jl_datatype_t *dt = (jl_datatype_t*)ty;
+
+    // Based on section 5.4 of the Procedure Call Standard
+
+    // VFP standard variant: see 6.1.2.2
+    //   Any result whose type would satisfy the conditions for a VFP CPRC is
+    //   returned in the appropriate number of consecutive VFP registers
+    //   starting with the lowest numbered register (s0, d0, q0).
+    classify_cprc(dt, vfpreg);
+    if (*vfpreg)
+        return;
+
+    // - A Half-precision Floating Point Type is returned in the least
+    //   significant 16 bits of r0.
+    if (dt == jl_float16_type) {
+        *reg = true;
+        return;
+    }
+
+    // - A Fundamental Data Type that is smaller than 4 bytes is zero- or
+    //   sign-extended to a word and returned in r0.
+    // - A double-word sized Fundamental Data Type (e.g., long long, double and
+    //   64-bit containerized vectors) is returned in r0 and r1.
+    // - A word-sized Fundamental Data Type (eg., int, float) is returned in r0.
+    // NOTE: assuming "fundamental type" == jl_is_bitstype, might need exact def
+    if (jl_is_bitstype(dt) && dt->size <= 8) {
+        *reg = true;
+        return;
+    }
+
+    // A 128-bit containerized vector is returned in r0-r3.
+    size_t vec_size = isContainerizedVector(dt);
+    if (vec_size == 128) {
```

Comment:
I won't worry about arm without vfp too much since IIRC they have other issues (don't pass number tests?). Leaving a comment here should probably be enough.